language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -121,28 +130,20 @@ instance ( PrettyBy config (Term TyName Name ())
         , "}"
         ]
 
-data ExTally = ExTally
-    { _exTallyCPU    :: ExTallyCounter ExCPU
-    , _exTallyMemory :: ExTallyCounter ExMemory
-    }
-    deriving stock (Eq, Generic, Show)
-    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
-instance ( PrettyBy config (Term TyName Name ())
-         ) => PrettyBy config ExTally where
-    prettyBy config (ExTally cpu memory) = parens $ fold
-        [ "{ cpu: ", prettyBy config cpu, line
-        , "| mem: ", prettyBy config memory, line
-        , "}"
-        ]
+-- | A human-readable string to determine which kind of operation caused the budget increase.
```

Comment:
I think we rather want this to be a big algebraic datatype of alternatives, e.g.
```
data Operation = ComputeTyInst | ... 
```
obviously we want to pretty-print it, but given we know all the cases, I can't see a reason to use a string rather than an explicit set of cases.

Reply:
What'd be the benefits? Kinda strange to argue for less types, but I don't see an argument either way, so Strings are less trouble.

Reply:
Well, this is also going to be part of the key for the cost model function, so all the usual benefits: ensures exhaustiveness, not missing cases, correctly updated, etc etc.

But I'd suggest it on the bare principle that we generally avoid strings, and especially if we have a known, exhaustive list of cases we should use a datatype.